File: sethv.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <time.h>
#include <sys/time.h>
#include <math.h>
#include <unistd.h>
#include <termios.h>


#include <wiringPi.h>
#include <wiringPiSPI.h>
#include <mcp23x0817.h>
#include <mcp23s17.h>
#include <softPwm.h>
#include <linux/spi/spidev.h>
#include <ad5685.h>

#define NSTEPS 100
#define MCPPINBASE 2000

#define SPICS 1

#define SPISPEED 40000000
//#define SPISPEED 320000

AD5685 dac[3];


int mygetch ( void ) 
{
  int ch;
  struct termios oldt, newt;

  tcgetattr ( STDIN_FILENO, &oldt );
  newt = oldt;
  newt.c_lflag &= ~( ICANON | ECHO );
  tcsetattr ( STDIN_FILENO, TCSANOW, &newt );
  ch = getchar();
  tcsetattr ( STDIN_FILENO, TCSANOW, &oldt );

  return ch;
}







void initialization(){

  
  wiringPiSetup () ;
  wiringPiSPISetup (SPICS, SPISPEED);


  //bring the MCP out of reset
  pinMode(26, OUTPUT);
  digitalWrite(26, HIGH);
  
  //setup MCP
  int retc = mcp23s17Setup (MCPPINBASE, SPICS, 0);
  printf("mcp setup done %d\n",retc);

  //sete RESET to DACs to high
  digitalWrite (MCPPINBASE+7, 1);
  pinMode(MCPPINBASE+7, OUTPUT);

  //sete LDAC to DACs to low
  digitalWrite (MCPPINBASE+3, 0);
  pinMode(MCPPINBASE+3, OUTPUT);


  
  AD5685_setup (&dac[0], MCPPINBASE, 4, MCPPINBASE, 2, MCPPINBASE, 0);
  AD5685_setup (&dac[1], MCPPINBASE, 5, MCPPINBASE, 2, MCPPINBASE, 0);
  AD5685_setup (&dac[2], MCPPINBASE, 6, MCPPINBASE, 2, MCPPINBASE, 0);  

}


int main(int argc, char *argv[])
{
	int opt;
	int cmderr = 0;

	/*
	while((opt = getopt(argc, argv, “:if:lrx”)) != -1) 
	  { 
	    switch(opt) 
	      { 
	      case ‘i’: 
	      case ‘l’: 
	      case ‘r’: 
                printf(“option: %c\n”, opt); 
                break; 
	      case ‘f’: 
                printf(“filename: %s\n”, optarg); 
                break; 
	      case ‘:’: 
                printf(“option needs a value\n”); 
                break; 
	      case ‘?’: 
                printf(“unknown option: %c\n”, optopt);
	      break; 
	      } 
	  } 
	*/
	initialization();

	int channel = atoi(argv[1]);
	float value = atof(argv[2]);

	value = value*2.3/1510.;
	int idac = (int) (channel/4);
	printf(" Chan %i HV idac %i  is set to %7.2f\n", channel, idac, value);	
	AD5685_setdac(&dac[idac],channel%4,value);

	
	
	return 0 ;
}   

File: sethv.py
from dac8164 import dac8164
from MCP23S08 import MCP23S08
import time
import sys


def initialization():
    dac = [None,None,None]

    # initialize MCP
    mcp1 = MCP23S08(bus=0x00, pin_cs=0x00, device_id=0x02)
    mcp1.open()
    mcp1._spi.max_speed_hz = 10000000

    dac[0] = dac8164(mcp1, 6, 7, 0, -1, -1)
    dac[1] = dac8164(mcp1, 3, 7, 0, -1, -1)
    dac[2] = dac8164(mcp1, 5, 7, 0, -1, -1)

    return dac



def set_hv(channel, value, dac):
    idac = int(channel/4)

    alphas = [0.9055, 0.9073, 0.9051, 0.9012, 0.9012, 0.9034,
              0.9009, 0.9027, 0.8977, 0.9012, 0.9015, 1]

    alpha = alphas[channel]

    digvalue = int(alpha*16383*value/2.5) & 0x3FFF

    dac[idac].DAC8164_writeChannel(channel, digvalue)


    

channel = int(sys.argv[1])
voltage = int(sys.argv[2])
nsteps = 200
dac = initialization()

value = voltage*2.3/1510

set_hv(channel,value,dac)

print("Channel " + str(channel) + " set to " + str(voltage) + " volts.")
    
